// eslint-plugin-jest/lib/rules/prefer-mock-return-shorthand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _utils = require("@typescript-eslint/utils");
var _utils2 = require("./utils");
const withOnce = (name, addOnce) => {
  return `${name}${addOnce ? 'Once' : ''}`;
};
const findSingleReturnArgumentNode = fnNode => {
  if (fnNode.body.type !== _utils.AST_NODE_TYPES.BlockStatement) {
    return fnNode.body;
  }
  if (fnNode.body.body[0]?.type === _utils.AST_NODE_TYPES.ReturnStatement) {
    return fnNode.body.body[0].argument;
  }
  return null;
};
var _default = exports.default = (0, _utils2.createRule)({
  name: __filename,
  meta: {
    docs: {
      description: 'Prefer mock return shorthands'
    },
    messages: {
      useMockShorthand: 'Prefer {{ replacement }}'
    },
    schema: [],
    type: 'suggestion',
    fixable: 'code'
  },
  defaultOptions: [],
  create(context) {
    return {
      CallExpression(node) {
        if (node.callee.type !== _utils.AST_NODE_TYPES.MemberExpression || !(0, _utils2.isSupportedAccessor)(node.callee.property) || node.arguments.length === 0) {
          return;
        }
        const {
          property
        } = node.callee;
        const mockFnName = (0, _utils2.getAccessorValue)(property);
        const isOnce = mockFnName.endsWith('Once');
        if (mockFnName !== withOnce('mockImplementation', isOnce)) {
          return;
        }
        const [arg] = node.arguments;
        if (!(0, _utils2.isFunction)(arg) || arg.params.length !== 0) {
          return;
        }
        const replacement = withOnce('mockReturnValue', isOnce);
        const returnNode = findSingleReturnArgumentNode(arg);
        if (!returnNode) {
          return;
        }
        context.report({
          node: property,
          messageId: 'useMockShorthand',
          data: {
            replacement
          },
          fix(fixer) {
            return [fixer.replaceText(property, replacement), fixer.replaceText(arg, context.sourceCode.getText(returnNode))];
          }
        });
      }
    };
  }
});
